ext/http/01_http.js

```diff
@@ -74,104 +75,82 @@ class HttpConn {
   #closed = false;
   #remoteAddr;
   #localAddr;
-
-  // This set holds resource ids of resources
-  // that were created during lifecycle of this request.
-  // When the connection is closed these resources should be closed
-  // as well.
-  managedResources = new Set();
+  #connections;
+  #waiters;
+  #httpRid;
+  #requests;
+  #error;
 
   constructor(rid, remoteAddr, localAddr) {
-    this.#rid = rid;
     this.#remoteAddr = remoteAddr;
     this.#localAddr = localAddr;
+    this.#connections = [];
+    this.#waiters = [];
+    this.#requests = new Set();
+
+    const connections = this.#connections;
+    const waiters = this.#waiters;
+    const context = new CallbackContext();
+    const onError = this.#onError.bind(this);
+
+    this.#httpRid = core.ops.op_serve_http_on(rid, (fallback_base_url) => context.initialize(fallback_base_url), map_to_userland(this.#requests, context, (req) => {
+      const promise = new Promise(r => {
+        connections.push([r, req]);
+        waiters.pop()?.(true);
+      });
+      return promise;
+    }, onError));
+    // Start the async IIFE to avoid looking like the event loop stalled
+    const httpRid = this.#httpRid;
+    (async () => {
+      try {
+        await core.opAsync("op_http_wait", httpRid);
+      } catch (e) {
+        onError(e);
+      }
+    })()
   }
 
-  /** @returns {number} */
-  get rid() {
-    return this.#rid;
+  #onError(e) {
+    // Wake any waiters so they throw the exception
+    this.#error = e;
+    for (const waiter of this.#waiters) {
+      waiter(true);
+    }           
   }
 
   /** @returns {Promise<RequestEvent | null>} */
   async nextRequest() {
-    let nextRequest;
-    try {
-      nextRequest = await core.opAsync("op_http_accept", this.#rid);
-    } catch (error) {
-      this.close();
-      // A connection error seen here would cause disrupted responses to throw
-      // a generic `BadResource` error. Instead store this error and replace
-      // those with it.
-      this[connErrorSymbol] = error;
-      if (
-        ObjectPrototypeIsPrototypeOf(BadResourcePrototype, error) ||
-        ObjectPrototypeIsPrototypeOf(InterruptedPrototype, error) ||
-        StringPrototypeIncludes(error.message, "connection closed")
-      ) {
+    while (true) {
+      if (this.#error) {
+        throw this.#error;
+      }
+      const conn = this.#connections.pop();
+      if (conn) {
+        return { request: conn[1], respondWith: (resp) => { 
+          conn[0](resp);
+          // TODO(mmastrac): How do we resolve this promise? Do we?
+          return new Promise(r => r());
+        } };
+      }
+      const wakeup = new Promise(r => this.#waiters.push(r));
+      if (await wakeup === false) {
```

Comment:
Could you explain why these wakeups are needed?

Reply:
This is the cleanest approach I could think of to translate the callback-style of new API to the older, async-style `serveHttp`. 

Each caller to `nextRequest` looks for a connection in the queue. If it finds one, it returns right away. If not, it  parks itself as a promise in `waiters`, and we wake one of them every time we get a new request, allowing one of them to grab the next request. If the waiter promise returns true, that means this `nextRequest` promise should grab a connection and resolve its promise. If it returns false, we're shutting down and it should return null.

It's possible we could simplify this and assert that there is never more than one `nextRequest` in-flight (I don't actually know if the existing API supports it - I need to write a test to confirm).